std/container/array.d

```diff
@@ -428,32 +481,28 @@ if (!is(immutable T == immutable bool))
     this(U)(U[] values...)
     if (isImplicitlyConvertible!(U, T))
     {
-        import core.lifetime : emplace;
-
-        static if (T.sizeof == 1)
-        {
-            const nbytes = values.length;
-        }
-        else
+        // [2021-07-17] Checking to see whether *always* calling ensureInitialized works-around-and/or-is-related-to https://issues.dlang.org/show_bug.cgihttps://issues.dlang.org/show_bug.cgi...
+        //if (values.length)
         {
-            import core.checkedint : mulu;
-            bool overflow;
-            const nbytes = mulu(values.length, T.sizeof, overflow);
-            if (overflow)
-                assert(false, "Overflow");
-        }
-        auto p = cast(T*) enforceMalloc(nbytes);
-        // Before it is added to the gc, initialize the newly allocated memory
-        foreach (i, e; values)
-        {
-            emplace(p + i, e);
-        }
-        static if (hasIndirections!T)
-        {
-            if (p)
-                GC.addRange(p, T.sizeof * values.length);
+            _data.refCountedStore.ensureInitialized();
+            _data.reserve(values.length);
+            foreach (ref value; values)
+            {
+                // We do not simply write "_data.insertBack(value);"
+                // because that might perform, on each iteration, a now-redundant check of length vs capacity.
+                // Thanks to @dkorpel (https://github.com/dlang/phobos/pull/8162#discussion_r667479090).
+
+                import core.lifetime : emplace;
+                emplace(_data._payload.ptr + _data._payload.length, value);
+
+                // We increment the length after each iteration (as opposed to adjusting it just once, after the loop)
+                // in order to improve error-safety (in case one of the calls to emplace throws).
```

Comment:
So if one of the calls to `emplace` throws, the `Array` will be partially populated, with some unspecified `length` <= `values.length`? This does not seem like desirable behavior, and as far as I can tell is not how it worked before. Please preserve the original behavior here.

Reply:
@pbkacus: First, I apologize for long delay. Second:

For this case -- when one of the calls to `emplace` throws -- after *some* (but not *all*) calls to emplace have succeeded:

- the pre-existing behaviour is that the `length` remains unchanged -- just as if *no* elements have been added -- but IMO this is incorrect because *some* elements have been successfully added;

- the proposed behaviour is that `length` is increased with the number of elements which have been successfully added (indeed, a number less than `values.length`);

- but `length` should not be increased with `values.length`, because that would signify that *all* requested elements have been successfully added.

Am I mis-understanding anything, please ?

Reply:
Your understanding is correct. However, I disagree with your opinion that the pre-existing behavior is wrong.

If the user requests an array containing a specific list of values, we must either give them what they requested, or report failure via an exception. Giving them *part* of what they requested is not acceptable. Users must be able to rely on the standard library to do what they ask it to do.

It is also highly inappropriate to introduce a behavioral change like this in a PR whose stated purpose is to "reduce code duplication."

Reply:
> 
> 
> Your understanding is correct. However, I disagree with your opinion that the pre-existing behavior is wrong.
> 
> If the user requests an array containing a specific list of values, we must either give them what they requested, or report failure via an exception. Giving them _part_ of what they requested is not acceptable. Users must be able to rely on the standard library to do what they ask it to do.

I agree that this constructor should either successfully construct an array containing all requested elements or propagate the Exception.

I believe that this is done both by pre-existing code and by the proposed changeset.

However !!

I believe that the pre-existing code incorrectly tracks the number of (successfully) constructed elements during the loop in the constructor. The tracked number of elements is zero during the entire loop and is only updated after the loop -- that is: if the entire loop has been successfully completed.

But if the loop is not completed:

- yes, the `Exception` is propagated (the user is not misled into thinking that his/her request for construction has completed successfully) -- this is good (and is kept in the proposed changeset);

- no, the already-constructed elements are not destroyed (because they have not been tracked) -- this is bad (and is corrected by the proposed changeset).

I have just added a `unittest`. With the pre-existing code, the `unittest` fails.

(The pre-existing code is temporarily enabled with `static if (1)` at line 553 in the current patchset, i.e. in https://github.com/DoctorNoobingstoneIPresume/dlang_phobos/commit/cd46b4ef7e8a165bc6bf56cc0c0e3511fb89541e. This is just temporary so we can easily test together. If we change that to "static if (0)", the new code is enabled and the `unittest` passes.)

> It is also highly inappropriate to introduce a behavioral change like this in a PR whose stated purpose is to "reduce code duplication."

I agree that the commit message (and possibly the code) can be improved. I have just tried to modify the commit message now. I hope it is better.

Reply:
> (The pre-existing code is temporarily enabled with `static if (1)` at line 553 in the current patchset, i.e. in [DoctorNoobingstoneIPresume@cd46b4e](https://github.com/DoctorNoobingstoneIPresume/dlang_phobos/commit/cd46b4ef7e8a165bc6bf56cc0c0e3511fb89541e). This is just temporary so we can easily test together: if we change that line to `static if (0)`, the new code is enabled and the `unittest` passes.)

I have just changed that line to `static if (0)` in order to enable the new code: https://github.com/DoctorNoobingstoneIPresume/dlang_phobos/commit/26ca49aeecb9d930adfe3849630348e3091bbc03.

BTW: In Github, is there a way to see all the commits that a proposed changeset has referred to during its evolution, please ?

Reply:
> I have just added a `unittest`. With the pre-existing code, the `unittest` fails.

The `unittest` can be found by searching for `S.s_nDestroyed == S.s_nConstructed`.

Reply:
Thanks for investigating this. It looks like this is a bug in the pre-existing code, so the best way to handle it would be to file an issue on [issues.dlang.org][1], and submit the fix as a separate PR.

[1]: https://issues.dlang.org/

Reply:
An issue has been added as https://issues.dlang.org/show_bug.cgi?id=22337.

Could you please let me know whether the commit message of the PR should be modified to specify the issue number ?

Regarding making a separate PR: In this case, I believe this is just extra work for little benefit. With the current PR, we get both a fixed bug and simpler-and-more-clear code (if we do not count the `unittest`'s and the comments). The simpler code (with fewer lines of code) helps us reason about the bug and the fix.

Reply:
The commit message for a bug fix should be formatted according to these instructions: <https://wiki.dlang.org/Contributing_to_Phobos#Issues>

Using separate PRs for bug fixes and refactors makes the changes easier to review and makes the git history of the code more granular (which makes it easier to `git bisect`, for example). In any case, it is D policy not to mix refactors and bug fixes; see here: <https://github.com/dlang/dmd/blob/master/CONTRIBUTING.md#refactoring>

Reply:
Is this not already very easy to review ?

Reply:
How easy a PR is to review is a subjective matter. However, we do have objective guidelines and those are that refactorings and bug fixes should go to separate PRs. I suggest that you use this PR to do the refactoring and make a subsequent PR to fix the bug.

This matter has been extensively discussed and contributors occasionally have arguments against this policy, however, experience has shown that it is best to separate refactorings from bug fixes.

Reply:
The "objective" guidelines referred above (https://github.com/dlang/dmd/blob/master/CONTRIBUTING.md#refactoring) state that:

"[...] Pull requests that do non-trivial refactorings should not include other changes, such as new feature or bug fix. [...]"

IMO, this "refactoring" is trivial.